rfd/0-job-worker.md

```diff
@@ -0,0 +1,240 @@
+---
+authors: Joshua Rubin (me@jawa.dev)
+state: draft
+---
+
+# RFD 0 - Prototype Job Worker Service
+
+## What
+
+Implement a prototype job worker service that provides an API to run arbitrary Linux processes.
+
+## Details
+
+### Design Approach
+
+Follow the requirements of [the challenge](https://github.com/gravitational/careers/blob/main/challenges/systems/challenge-1.md) as closely as possible and do not unnecessarily expand the scope.
+
+### Proposed API
+
+The gRPC API is defined in `jobworker.proto` and the worker library interface is defined in `worker.go`.
+
+#### Worker Library
+
+The worker library defines a simple interface with methods to start, stop, query status and get the output of a job. The implementation is responsible for starting jobs, keeping track of their output, status and exit codes.
+
+##### cgroups
+
+The library implementation will be configured at instantiation with limits that can be parsed into values for `cpu.max`, `memory.max` and `io.max`. These values will originate from runtime flags provided to the server with useful defaults. All processes will be started in a new cgroup regardless of the values of these limits. Note that for expediency, it is assumed that cgroups v2 are in use on the system running the service.
+
+##### namespaces
+
+The library implementation will be configured such that it knows how to execute the server binary as a child process, with the proper clone and unshare flags to execute in a new pid, mount and network namespace, and call back into the library to execute the given command for the job. This library instance is obviously separate from the main server process, but due to being reexecuted with the same flags, just a different command (`child` instead of `serve`, see CLI UX below), it will retain all the necessary configuration.
+
+##### Job Execution
+
+The server will initially reexecute itself, using `exec.Command()`, as follows:
+
+```sh
+/proc/self/exe child [serve flags] -- command [args]...
+```
+
+When calling this, it will set the command's `SysProcAttr` with:
+
+```go
+syscall.SysProcAttr{
+    Cloneflags: syscall.CLONE_NEWPID | // New pid namespace
+        syscall.CLONE_NEWNS | // New mount namespace group
+        syscall.CLONE_NEWNET, // New network namespace
```

Comment:
This is more of an hypothetical, but how could we make the NET namespace more useful?

Reply:
ah, i see, this is the _real_ interview isn't it? 🤣 
well, we could create a bridge device for the worker and virtual ethernet devices for each job. we'd connect the bridge to the physical and virtual devices. we could give the virtual devices their own cidr range. that would make each job (don't call them containers) independently addressable from the host.

Reply:
Sounds good!

> ah, i see, this is the real interview isn't it? 🤣

😇